ndonnx/extensions.py

```diff
@@ -0,0 +1,437 @@
+    xty = x._tyarray
+    if isinstance(xty, tydx.masked_onnx.TyMaArray):
+        result_type = ndx.result_type(xty.data.dtype, value_.dtype)
+        if xty.mask is None:
+            res: tydx.TyArrayBase = xty.data
+        else:
+            res = tydx.funcs.where(xty.mask, value_, xty.data)
+    elif isinstance(xty, tydx.onnx.TyArray):
+        result_type = ndx.result_type(xty.dtype, value_.dtype)
+        res = xty.astype(result_type)
+    else:
+        raise TypeError("'fill_null' is only implemented for built-in types")
+
+    return ndx.Array._from_tyarray(res).astype(result_type)
+
+
+def make_nullable(
+    x: ndx.Array,
+    null: ndx.Array | None,
+    /,
+    *,
+    merge_strategy: Literal["raise", "merge"] = "raise",
+) -> ndx.Array:
+    """Given an array ``x`` of values and a null mask ``null``, construct a new Array
+    with a nullable data type.
+
+    Parameters
+    ----------
+    x: Array
+        Array of values
+
+    null: Array
+        Array of booleans indicating whether each element of ``x`` is null.
+
+    merge_strategy: Literal["raise", "merge"]
+        If `"raise"`, a ``TypeError`` is raised if ``x`` is already of
+        a nullable data type. If `"merge"` is provided, any mask
+        existing on ``x`` is merged with ``null``.
+
+    Returns
+    -------
+    out: Array
+        A new Array with a nullable data type.
+
+    Raises
+    ------
+    TypeError
+        If the data type of ``x`` does not have a nullable counterpart.
+    """
+    x = x.copy()
+    null = None if null is None else null.copy()
+
+    if null is None:
+        if isinstance(x._tyarray, TyMaArray):
+            return x
+        if isinstance(x._tyarray, tydx.onnx.TyArray):
+            return ndx.Array._from_tyarray(
+                tydx.masked_onnx.make_nullable(x._tyarray, None)
+            )
+        raise TypeError(f"failed to make array of data type `{x.dtype}` nullable")
+    if not isinstance(null._tyarray, tydx.onnx.TyArrayBool):
+        raise TypeError(f"'null' must be of boolean data type, found `{null.dtype}`")
+    if isinstance(x._tyarray, tydx.onnx.TyArray):
+        return ndx.Array._from_tyarray(
+            tydx.masked_onnx.make_nullable(x._tyarray, null._tyarray)
+        )
+    if merge_strategy == "merge" and isinstance(x._tyarray, tydx.masked_onnx.TyMaArray):
+        mask = tydx.masked_onnx.merge_masks(x._tyarray.mask, null._tyarray)
+        tyarr = tydx.masked_onnx.make_nullable(x._tyarray.data, mask)
+        return ndx.Array._from_tyarray(tyarr)
+    if isinstance(x._tyarray, tydx.datetime.TimeBaseArray):
+        # TODO: The semantics of this branch are very odd!
+        is_nat = x._tyarray.is_nat
+        merged = tydx.masked_onnx.merge_masks(is_nat, null._tyarray)
+        if merged is not None:
+            is_nat = merged
+        return ndx.Array._from_tyarray(x._tyarray.dtype._build(data=x._tyarray._data))
+
+    raise ndx.UnsupportedOperationError(
```

Comment:
You have deprecated this error but are raising it here. Does this not appear in the test suite?

Reply:
There are many deprecation warnings raised by the tests suite at the moment. I'd rather fix them in a follow-up PR. Functionally, `ndx.UnsupportedOperationError` is an alias for `TypeError` now. Nonetheless, it shouldn't be used here.